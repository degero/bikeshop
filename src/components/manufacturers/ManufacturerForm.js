import React, { useEffect } from "react";
import { useForm } from "react-hook-form";


const ManufacturerForm = ({ manufacturer, onSave, saving = false }) => {
  const { register, handleSubmit, setValue, formState: { errors } } = useForm(
    { defaultValues: { name: manufacturer.name } }
  );
  let formWrapperStyle = "form-group";

  if (errors && errors.length > 0) {
    formWrapperStyle += " has-error";
  }

  const submit = data => {
    onSave(data);
  }

  const errorSection = (errorField, type, message) => {
    if (errorField?.type === type) {
      return (
        <div className="alert alert-danger" role="alert">
          <p className="error-text">{message ?? errorField.message}</p>
        </div>
      )
    }
  }
  useEffect(() => {
    if (manufacturer.id != null) {
      // edit form set values
      const fields = ['name'];
      fields.forEach(field => setValue(field, manufacturer[field]));
    }
  }, [manufacturer, setValue]);

  return (
    <div className="row content">
      <div className="col-12">
        <h2>{manufacturer.id ? "Edit" : "Create"} Manufacturer</h2>
        <p>* using react-hook-form</p>
      </div>
      <div className="col-12 col-md-4">
        <form onSubmit={handleSubmit(submit)} autoComplete="off">
          <div className={formWrapperStyle}>
            <label htmlFor="name">Manufacturer name:</label>
            <input className="form-control" type="text" id="name"
              {...register("name", { required: "Manufacturer name is required" })}
            />
          </div>
          {errorSection(errors?.name, "required")}
          <button
            className="btn btn-primary"
            type="submit"
          >{saving ? "Saving" : "Save"}</button>
        </form>
      </div>
    </div>
  );
}


export default ManufacturerForm;
