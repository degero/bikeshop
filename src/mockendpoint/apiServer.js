/* json-server for dev use, see .env.development api endpoint setting */
const jsonServer = require("json-server");
const server = jsonServer.create();
const path = require("path");
const router = jsonServer.router(path.join(__dirname, "db.json"));

const middleware = jsonServer.defaults();
server.use(middleware);

server.use(jsonServer.bodyParser);

// Slow call down 1500ms
server.use(function (req, res, next) {
  setTimeout(next, 1500);
});

server.use((req, res, next) => {
  // set a create date on post
  if (req.method === "POST") {
    req.body.createdAt = Date.now();
  }
  // TODO add updated date on PUT
  next();
});

server.post("/bikes/", function (req, res, next) {
  const error = validateBike(req.body);
  if (error) {
    res.status(400).send(error);
  } else {
    req.body.slug = createSlug(req.body.manufacturer + " " + req.body.model); // Generate a slug for new bike.
    next();
  }
});

server.use(router);

// App runs on 3000 and api on 3001
const port = 3001;
server.listen(port, () => {
  console.log(`JSON Server is running on port ${port}`);
});

function createSlug(value) {
  return value
    .replace(/[^a-z0-9_]+/gi, "-")
    .replace(/^-|-$/g, "")
    .toLowerCase();
}

// Server side validation of bike
function validateBike(bike) {
  if (!bike.manufacturer) return "Manufacturer is required.";
  if (!bike.model) return "Model is required.";
  if (!bike.price) return "Price is required.";
  return "";
}
