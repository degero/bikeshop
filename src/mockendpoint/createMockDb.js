/* eslint-disable no-console */
const fs = require("fs");
const path = require("path");
const mockData = require("./mockData");

const { bikes } = mockData;
const data = JSON.stringify({ bikes });
const filepath = path.join(__dirname, "db.json");

fs.writeFile(filepath, data, function (err) {
  err ? console.log(err) : console.log("Mock DB db.json created.");
});
