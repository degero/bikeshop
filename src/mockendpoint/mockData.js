const bikes = [
  {
    id: 1,
    manufacturer: "Santa Cruz",
    model: "Bronson",
    price: 6999,
    slug: "santa-cruz-bronson",
  },
];

const manufacturers = [
  {
    id: 1,
    name: "Santa Cruz",
    slug: "santa-cruz"
  },
];

module.exports = {
  bikes,
  manufacturers
};
