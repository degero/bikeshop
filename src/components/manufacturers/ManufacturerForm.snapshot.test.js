import React from "react";
import renderer from "react-test-renderer";
import ManufacturerForm from "./ManufacturerForm";
import { manufacturers } from "../../mockendpoint/mockData";

describe("ManufacturerForm component - snapshots", () => {
  it("sets submit button label 'Saving...' when saving is true", () => {
    // ARRANGE / ACT
    const tree = renderer.create(
      <ManufacturerForm
        manufacturer={manufacturers[0]}
        manufacturers={manufacturers}
        onSave={jest.fn()}
        onChange={jest.fn()}
        saving={true}
      />
    );
    // ASSERT
    expect(tree).toMatchSnapshot();
  });

  it("sets submit button label 'Save' when saving is false", () => {
    // ARRANGE / ACT
    const tree = renderer.create(
      <ManufacturerForm
        manufacturer={manufacturers[0]}
        manufacturers={manufacturers}
        onSave={jest.fn()}
        onChange={jest.fn()}
        saving={false}
      />
    );
    // ASSERT
    expect(tree).toMatchSnapshot();
  });
});
