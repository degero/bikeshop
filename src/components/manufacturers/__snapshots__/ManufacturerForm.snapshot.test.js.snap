// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ManufacturerForm component - snapshots sets submit button label 'Save' when saving is false 1`] = `
<div
  className="row content"
>
  <div
    className="col-12"
  >
    <h2>
      Edit
       Manufacturer
    </h2>
    <p>
      * using react-hook-form
    </p>
  </div>
  <div
    className="col-12 col-md-4"
  >
    <form
      autoComplete="off"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="name"
        >
          Manufacturer name:
        </label>
        <input
          className="form-control"
          id="name"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
        />
      </div>
      <button
        className="btn btn-primary"
        type="submit"
      >
        Save
      </button>
    </form>
  </div>
</div>
`;

exports[`ManufacturerForm component - snapshots sets submit button label 'Saving...' when saving is true 1`] = `
<div
  className="row content"
>
  <div
    className="col-12"
  >
    <h2>
      Edit
       Manufacturer
    </h2>
    <p>
      * using react-hook-form
    </p>
  </div>
  <div
    className="col-12 col-md-4"
  >
    <form
      autoComplete="off"
      onSubmit={[Function]}
    >
      <div
        className="form-group"
      >
        <label
          htmlFor="name"
        >
          Manufacturer name:
        </label>
        <input
          className="form-control"
          id="name"
          name="name"
          onBlur={[Function]}
          onChange={[Function]}
          type="text"
        />
      </div>
      <button
        className="btn btn-primary"
        type="submit"
      >
        Saving
      </button>
    </form>
  </div>
</div>
`;
